File: tutorial/sc17/bagoftasks-noft/errh_blank.f
C
C  Copyright (c) 2013-2017 The University of Tennessee and The University
C                          of Tennessee Research Foundation.  All rights
C                          reserved.
C  $COPYRIGHT$
C
C  Additional copyrights may follow
C
C  $HEADER$
C

C     Simulate the BLANK mode in FTMPI. Only peers communicating
C     directly with a dead process will notice the failures, and no
C     corrective actions will be taken globally (in other terms
C     communicator revocation is not an option). While this approach can
C     minimize the overhead, it makes difficult to recover the failure
C     of the master.

      subroutine error_handler(communicator, error_code)

C     Only the master will react to the failure. Every other process
C     will ignore all MPI_ERR_PROC_FAILED except if the dead process is
C     the master.

      implicit none
      include "mpif.h"
      include "fsolvergen.inc"

C input Args
      integer communicator, error_code !, stdargs(*)

C Local Variables
      integer rc, myrank, num_fails
C     Failed group handle and members array
      integer f_group, f_group_rank(maxworker)
C     Original group handle and members array
      integer w_group, w_group_rank(maxworkers)

c     Who I was on the original communicator (for debugging purposes)
      call MPI_Comm_rank(communicator, myrank, rc)

c     A failed process: verify the error code
      if()
c     Get failed procs group
c     Get no. of failed procs
c     Get group from current communicator
c     Get ranks of failed procs
c     And mark them dead
            do rc = 1, num_fails
               write(*,*) "[", myrank, "] Peer",
     $              w_group_rank(rc), "dead"
               call mark_dead(w_group_rank(rc))
            enddo
            rc = 0
         endif
c     Strange things happen
      else
         write(*,*)"Something weird is happening.",
     $        "Error code", error_code, "on rank", myrank
      endif

      return
      end

File: tutorial/bagoftasks/c/answer/errh_blank.c
/* 
 *   Copyright (c) 2013-2021 The University of Tennessee and The University
 *                           of Tennessee Research Foundation.  All rights
 *                           reserved.
 *   $COPYRIGHT$
 * 
 *   Additional copyrights may follow
 * 
 *   $HEADER$
 * 

 *      Simulate the BLANK mode in FTMPI. Only peers communicating
 *      directly with a dead process will notice the failures, and no
 *      corrective actions will be taken globally (in other terms
 *      communicator revocation is not an option). While this approach can
 *      minimize the overhead, it makes difficult to recover the failure
 *      of the master.
 */

#include <stdio.h>
#include <mpi.h>
#include <mpi-ext.h>
#include "fsolvergen.h"

/*      Only the master will react to the failure. Every other process
 *      will ignore all MPI_ERR_PROC_FAILED except if the dead process is
 *      the master. */

void error_handler(MPI_Comm *pcommunicator, int *error_code, ...) {
   // locals
   MPI_Comm communicator = *pcommunicator;
   int rc, i, myrank, num_fails, error_class;
   char error_string[MPI_MAX_ERROR_STRING];
   // Failed group handle and members array
   MPI_Group f_group;
   int f_group_rank[MAX_SIZE];
   // Original group handle and members array
   MPI_Group w_group;
   int w_group_rank[MAX_SIZE];

   // Who I was on the original communicator (for debugging purposes)
   MPI_Comm_rank(communicator, &myrank);
   // user-friendly error string
   MPI_Error_string(*error_code, error_string, &i);

   printf("R%02d: Errhandler triggered %d: %s\n", myrank, *error_code, error_string);
   // A failed process: verify the error code
   MPI_Error_class(*error_code, &error_class);
   if(MPI_ERR_PROC_FAILED == error_class) {
      if(0 == myrank) {
         //     Get group from current communicator
         MPI_Comm_group(communicator, &w_group);
         //     Get failed procs group
         MPIX_Comm_failure_ack(communicator);
         MPIX_Comm_failure_get_acked(communicator, &f_group);
         //     Get no. of failed procs
         MPI_Group_size(f_group, &num_fails);
         //     Get ranks of failed procs in the original communicator
         for(i = 0; i < num_fails; i++) f_group_rank[i] = i;
         MPI_Group_translate_ranks(f_group, num_fails, f_group_rank,
                                   w_group, w_group_rank);
         //     And mark them dead
         for(i = 0; i < num_fails; i++) {
            printf("R%02d: Peer R%02d dead\n", myrank, w_group_rank[i]);
            mark_dead(w_group_rank[i]);
         }
         return;
      }
   }
   //     Strange things happen: master dead?
   printf("R%02d: Something weird is happening, aborting...\n", myrank);
   MPI_Abort(MPI_COMM_SELF, 2);
}
